File: android/app/src/main/java/com/cavadvpn/crypto/NoiseHandshake.kt
package com.cavadvpn.crypto

import java.security.MessageDigest
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/** Noise protocol identifier. */
const val PROTOCOL_NAME = "Noise_XX_25519_ChaChaPoly_SHA256"

/**
 * Post-handshake cipher state using ChaCha20-Poly1305 with incrementing nonce counter.
 * Nonce: 4 zero bytes + 8-byte little-endian counter.
 *
 * The ChaCha20Poly1305 cipher is created ONCE and re-initialized per packet
 * (avoids object allocation on every encrypt/decrypt — same fix as Go/Python).
 */
class NoiseCipherState(key: ByteArray) {
    private val key: ByteArray = key.copyOf(KEY_SIZE)
    private var counter: Long = 0L
    // Cipher created once; re-initialized with new nonce on each call.
    private val cipher = org.bouncycastle.crypto.modes.ChaCha20Poly1305()
    private val nonceBuf = ByteArray(NONCE_SIZE) // reused, avoids allocation per packet

    fun encrypt(plaintext: ByteArray, aad: ByteArray = ByteArray(0)): ByteArray {
        buildNonce(counter++, nonceBuf)
        return encryptWithCipher(cipher, key, nonceBuf, plaintext, aad)
    }

    fun decrypt(ciphertext: ByteArray, aad: ByteArray = ByteArray(0)): ByteArray {
        buildNonce(counter++, nonceBuf)
        return decryptWithCipher(cipher, key, nonceBuf, ciphertext, aad)
    }

    private fun buildNonce(n: Long, buf: ByteArray) {
        // 4 zero bytes + 8-byte little-endian counter
        buf[0] = 0; buf[1] = 0; buf[2] = 0; buf[3] = 0
        buf[4]  = (n         ).toByte()
        buf[5]  = (n ushr  8 ).toByte()
        buf[6]  = (n ushr 16 ).toByte()
        buf[7]  = (n ushr 24 ).toByte()
        buf[8]  = (n ushr 32 ).toByte()
        buf[9]  = (n ushr 40 ).toByte()
        buf[10] = (n ushr 48 ).toByte()
        buf[11] = (n ushr 56 ).toByte()
    }
}

/** Noise HKDF using HMAC-SHA256. Returns n (2 or 3) 32-byte outputs. */
internal fun noiseHKDF(ck: ByteArray, ikm: ByteArray, n: Int): List<ByteArray> {
    require(n in 2..3) { "noiseHKDF: n must be 2 or 3" }
    fun hmac(key: ByteArray, data: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(data)
    }
    val tempKey = hmac(ck, ikm)
    val o1 = hmac(tempKey, byteArrayOf(0x01))
    val o2 = hmac(tempKey, o1 + byteArrayOf(0x02))
    if (n == 2) return listOf(o1, o2)
    val o3 = hmac(tempKey, o2 + byteArrayOf(0x03))
    return listOf(o1, o2, o3)
}

/** Symmetric state tracking transcript hash and chaining key. */
internal class NoiseSymmetricState {
    private var ck: ByteArray
    private var h: ByteArray
    private var cs: NoiseCipherState? = null

    init {
        // Per Noise spec: if len(protocol_name) <= HASHLEN use raw bytes, else hash it.
        // PROTOCOL_NAME is exactly 32 bytes so we copy directly.
        val nameBytes = PROTOCOL_NAME.toByteArray(Charsets.UTF_8)
        val initial = if (nameBytes.size <= KEY_SIZE) {
            ByteArray(KEY_SIZE).also { System.arraycopy(nameBytes, 0, it, 0, nameBytes.size) }
        } else {
            MessageDigest.getInstance("SHA-256").digest(nameBytes)
        }
        ck = initial.copyOf()
        h  = initial.copyOf()
    }

    fun mixHash(data: ByteArray) {
        val d = MessageDigest.getInstance("SHA-256")
        d.update(h)
        d.update(data)
        h = d.digest()
    }

    fun mixKey(ikm: ByteArray) {
        val (newCk, k) = noiseHKDF(ck, ikm, 2)
        ck = newCk
        cs = NoiseCipherState(k)
    }

    fun encryptAndHash(plaintext: ByteArray): ByteArray {
        val ct = cs?.encrypt(plaintext, h) ?: plaintext
        mixHash(ct)
        return ct
    }

    fun decryptAndHash(ciphertext: ByteArray): ByteArray {
        val pt = cs?.decrypt(ciphertext, h) ?: ciphertext
        mixHash(ciphertext)
        return pt
    }

    fun split(): Pair<NoiseCipherState, NoiseCipherState> {
        val (k1, k2) = noiseHKDF(ck, ByteArray(0), 2)
        return Pair(NoiseCipherState(k1), NoiseCipherState(k2))
    }
}

/**
 * Result of a completed Noise_XX handshake.
 *
 * @property sendCipher for encrypting outgoing messages
 * @property recvCipher for decrypting incoming messages
 * @property remoteStatic 32-byte public key of the authenticated peer
 */
data class NoiseSession(
    val sendCipher: NoiseCipherState,
    val recvCipher: NoiseCipherState,
    val remoteStatic: ByteArray
) {
    override fun equals(other: Any?) = other is NoiseSession &&
        remoteStatic.contentEquals(other.remoteStatic)
    override fun hashCode() = remoteStatic.contentHashCode()
}

/**
 * Noise_XX initiator handshake (client side).
 *
 * Pattern:  -> e  |  <- e, ee, s, es  |  -> s, se
 */
class NoiseHandshake(private val staticKP: KeyPair) {
    private val ss = NoiseSymmetricState()
    private var e: KeyPair? = null
    private var re: ByteArray? = null
    private var rs: ByteArray? = null
    private var step = 0

    init { ss.mixHash(ByteArray(0)) } // empty prologue

    /** -> e: generate ephemeral key, returns 32-byte public key. */
    fun writeMessage1(): ByteArray {
        require(step == 0) { "writeMessage1: invalid step $step" }
        val eph = generateKeyPair()
        e = eph
        ss.mixHash(eph.publicKey)
        step = 1
        return eph.publicKey.copyOf()
    }

    /**
     * <- e, ee, s, es: process 80-byte server message.
     * @param msg re_pub(32) + EncryptAndHash(s)(48)
     */
    fun readMessage2(msg: ByteArray) {
        require(step == 1) { "readMessage2: invalid step $step" }
        val minLen = KEY_SIZE + KEY_SIZE + AEAD_OVERHEAD
        require(msg.size >= minLen) { "readMessage2: too short ${msg.size} < $minLen" }

        val re_ = msg.copyOf(KEY_SIZE)
        re = re_
        ss.mixHash(re_)

        val ee = diffieHellman(e!!.privateKey, re_)
        ss.mixKey(ee)

        val encStatic = msg.copyOfRange(KEY_SIZE, KEY_SIZE + KEY_SIZE + AEAD_OVERHEAD)
        rs = ss.decryptAndHash(encStatic)

        val es = diffieHellman(e!!.privateKey, rs!!)
        ss.mixKey(es)
        step = 2
    }

    /**
     * -> s, se: encrypt our static key, DH mix, split.
     * @return Pair(48-byte message, NoiseSession with send/recv ciphers)
     */
    fun writeMessage3(): Pair<ByteArray, NoiseSession> {
        require(step == 2) { "writeMessage3: invalid step $step" }

        val encStatic = ss.encryptAndHash(staticKP.publicKey)

        val se = diffieHellman(staticKP.privateKey, re!!)
        ss.mixKey(se)

        val (sendCs, recvCs) = ss.split()
        step = 3
        return Pair(encStatic, NoiseSession(sendCs, recvCs, rs!!.copyOf()))
    }
}

File: ios/Sources/CavadVPNCrypto/NoiseHandshake.swift
// NoiseHandshake.swift — Noise_XX initiator handshake (wire-compatible with Go server)
//
// Pattern:
//   -> e              (Message 1: 32 bytes)
//   <- e, ee, s, es   (Message 2: 80 bytes)
//   -> s, se          (Message 3: 48 bytes)

import Foundation
import Crypto

// MARK: - Protocol constants

private let protocolName = "Noise_XX_25519_ChaChaPoly_SHA256"

// MARK: - NoiseSymmetricState

/// Tracks the chaining key and transcript hash during a Noise handshake.
final class NoiseSymmetricState {
    private(set) var h: Data    // transcript hash
    private(set) var ck: Data   // chaining key
    private var cs: NoiseCipherState

    init() {
        let nameBytes = Data(protocolName.utf8)
        // Per Noise spec: if len(name) <= HASHLEN, pad with zeros; else SHA-256.
        let initial: Data
        if nameBytes.count <= keySize {
            var padded = Data(repeating: 0, count: keySize)
            padded.replaceSubrange(0..<nameBytes.count, with: nameBytes)
            initial = padded
        } else {
            initial = Data(SHA256.hash(data: nameBytes))
        }
        h  = initial
        ck = initial
        cs = NoiseCipherState()
    }

    func mixHash(_ data: Data) {
        var hasher = SHA256()
        hasher.update(data: h)
        hasher.update(data: data)
        h = Data(hasher.finalize())
    }

    func mixKey(_ ikm: Data) {
        let outputs = noiseHKDF(ck: ck, ikm: ikm, n: 2)
        ck = outputs[0]
        cs = NoiseCipherState()
        cs.initializeKey(outputs[1])
    }

    func encryptAndHash(_ plaintext: Data) throws -> Data {
        let ciphertext = try cs.hasKey
            ? cs.encrypt(plaintext: plaintext, aad: h)
            : plaintext
        mixHash(ciphertext)
        return ciphertext
    }

    func decryptAndHash(_ ciphertext: Data) throws -> Data {
        let plaintext = try cs.hasKey
            ? cs.decrypt(ciphertext: ciphertext, aad: h)
            : ciphertext
        mixHash(ciphertext)
        return plaintext
    }

    func split() -> (NoiseCipherState, NoiseCipherState) {
        let outputs = noiseHKDF(ck: ck, ikm: Data(), n: 2)
        let c1 = NoiseCipherState()
        c1.initializeKey(outputs[0])
        let c2 = NoiseCipherState()
        c2.initializeKey(outputs[1])
        return (c1, c2)
    }
}

// MARK: - NoiseSession

/// Result of a completed Noise_XX handshake.
public struct NoiseSession {
    public let sendCipher: NoiseCipherState
    public let recvCipher: NoiseCipherState
    public let remoteStatic: Data  // 32-byte peer public key

    init(sendCipher: NoiseCipherState, recvCipher: NoiseCipherState, remoteStatic: Data) {
        self.sendCipher   = sendCipher
        self.recvCipher   = recvCipher
        self.remoteStatic = remoteStatic
    }
}

// MARK: - NoiseHandshake

/// Noise_XX initiator (client-side) handshake.
public final class NoiseHandshake {
    private let staticKP: KeyPair
    private let ss = NoiseSymmetricState()
    private var ephemeralKP: KeyPair?
    private var remoteEphemeral: Data?
    private var remoteStatic: Data?
    private var step = 0

    public init(staticKP: KeyPair) {
        self.staticKP = staticKP
        ss.mixHash(Data()) // empty prologue
    }

    // MARK: Message 1: -> e

    /// Generates an ephemeral key pair, mixes the public key into h, and returns the 32-byte public key.
    public func writeMessage1() throws -> Data {
        guard step == 0 else { throw NoiseError.invalidStep(step) }
        let eph = generateKeyPair()
        ephemeralKP = eph
        ss.mixHash(eph.publicKey)
        step = 1
        return eph.publicKey
    }

    // MARK: Message 2: <- e, ee, s, es

    /// Processes the 80-byte server message: re.pub(32) + EncryptAndHash(s)(48).
    public func readMessage2(_ msg: Data) throws {
        guard step == 1 else { throw NoiseError.invalidStep(step) }
        let minLen = keySize + keySize + aeadOverhead   // 32 + 32 + 16 = 80
        guard msg.count >= minLen else { throw NoiseError.messageTooShort(msg.count, minLen) }
        guard let eph = ephemeralKP else { throw NoiseError.internalError("no ephemeral key") }

        // <- e
        let re = msg.prefix(keySize)
        remoteEphemeral = re
        ss.mixHash(re)

        // ee = DH(e_init, e_resp)
        let ee = try diffieHellman(privateKey: eph.privateKey, publicKey: Data(re))
        ss.mixKey(ee)

        // s: decrypt remote static
        let encS = msg[keySize ..< keySize + keySize + aeadOverhead]
        let rs = try ss.decryptAndHash(Data(encS))
        remoteStatic = rs

        // es = DH(e_init, s_resp)
        let es = try diffieHellman(privateKey: eph.privateKey, publicKey: rs)
        ss.mixKey(es)

        step = 2
    }

    // MARK: Message 3: -> s, se

    /// Encrypts our static public key, does se DH mix, splits into send/recv ciphers.
    /// - Returns: `(48-byte message, NoiseSession)`
    public func writeMessage3() throws -> (Data, NoiseSession) {
        guard step == 2 else { throw NoiseError.invalidStep(step) }
        guard let re = remoteEphemeral else { throw NoiseError.internalError("no remote ephemeral") }
        guard let rs = remoteStatic    else { throw NoiseError.internalError("no remote static") }

        // s: encrypt our static public key
        let encS = try ss.encryptAndHash(staticKP.publicKey)

        // se = DH(s_init, e_resp)
        let se = try diffieHellman(privateKey: staticKP.privateKey, publicKey: re)
        ss.mixKey(se)

        // split
        let (sendCs, recvCs) = ss.split()
        step = 3
        return (encS, NoiseSession(sendCipher: sendCs, recvCipher: recvCs, remoteStatic: rs))
    }
}

// MARK: - Errors

public enum NoiseError: Error {
    case invalidStep(Int)
    case messageTooShort(Int, Int)
    case internalError(String)
}
